test/pool-cl/helpers/CLCustomCurveHook.sol

```diff
@@ -0,0 +1,230 @@
+import {Constants} from "./Constants.sol";
+import {CLPoolParametersHelper} from "../../../src/pool-cl/libraries/CLPoolParametersHelper.sol";
+import {console2} from "forge-std/console2.sol";
+
+contract CLCustomCurveHook is BaseCLTestHook {
+    using CLPoolParametersHelper for bytes32;
+    using SafeCast for *;
+    using CurrencySettlement for Currency;
+    using BeforeSwapDeltaLibrary for BeforeSwapDelta;
+
+    IVault public immutable vault;
+    ICLPoolManager public immutable manager;
+
+    constructor(IVault _vault, ICLPoolManager _manager) {
+        vault = _vault;
+        manager = _manager;
+    }
+
+    // *****  Custom Curve Interface  *****//
+
+    function addLiquidity(PoolKey calldata key, uint256 amount0, uint256 amount1) public {
+        // 1. custom curve minting position token logic
+
+        // 2. deposit token into vault
+        vault.lock(abi.encode("deposit", abi.encode(key, amount0, amount1)));
+    }
+
+    function removeLiquidity(PoolKey calldata key, uint256 amount0, uint256 amount1, address recipient) public {
+        // 1. custom curve burning position token logic
+
+        // 2. withdraw token from vault
+        vault.lock(abi.encode("withdraw", abi.encode(key, amount0, amount1, recipient)));
+    }
+
+    // *****  ICLHooks Interface  *****//
+
+    function getHooksRegistrationBitmap() external pure override returns (uint16) {
+        return _hooksRegistrationBitmapFrom(
+            Permissions({
+                afterInitialize: true,
+                beforeAddLiquidity: true,
+                beforeSwap: true,
+                befreSwapReturnsDelta: true,
+                // disable all other hooks
+                beforeInitialize: false,
+                afterAddLiquidity: false,
+                beforeRemoveLiquidity: false,
+                afterRemoveLiquidity: false,
+                afterSwap: false,
+                beforeDonate: false,
+                afterDonate: false,
+                afterSwapReturnsDelta: false,
+                afterAddLiquidityReturnsDelta: false,
+                afterRemoveLiquidityReturnsDelta: false
+            })
+        );
+    }
+
+    function afterInitialize(address sender, PoolKey calldata key, uint160 sqrtPriceX96, int24 tick)
+        external
+        override
+        returns (bytes4)
+    {
+        require(sqrtPriceX96 == Constants.SQRT_RATIO_1_1, "InvalidSqrtPrice");
+        require(tick == 0, "InvalidTick");
+        require(key.parameters.getTickSpacing() == 10, "InvalidTickSpacing");
+
+        vault.lock(abi.encode("init", abi.encode(key)));
+
+        return ICLHooks.afterInitialize.selector;
+    }
+
+    function beforeAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        revert("banned adding liquidity by non-hook");
```

Comment:
This should be banned anyway even not taking this approach because custom curve will have its own liquidity logic